readme.md
[![Build status](https://ci.appveyor.com/api/projects/status/k6sdi0ym8cby98xm/branch/dev?svg=true)](https://ci.appveyor.com/project/gabdube/native-windows-gui/branch/dev)


# Native Windows GUI

<b>Native Windows GUI (NWG)</b> is a thin GUI toolkit built over the <b>Microsoft Windows WINAPI</b> for rust. The 
current version is <b>0.2.0 BETA 1</b>. The library is close to be production ready, but still lacks
some important features and some useful controls and resources.

NWG uses [retep998/winapi-rs](https://github.com/retep998/winapi-rs) and works on all rust channels and most
rust versions. NWG was tested on Windows 8.1 and Windows 10 using the MSVC ABI build but any version of Microsoft Windows supported by Rust is supposed to be
supported by NWG (vista and up).

## Why NWG?

Is native-windows-gui the gui framework you are looking for? It is ...

* For those who wants to develop on Windows and want the smallest executable and memory footprint possible. 
* For those who don't like dependencies. NWG only requires some `winapi-rs` crates and do not depends on external "executable" code 
* For those who want a canvas to draw pretty things, NWG has a very powerful (and light) canvas build over Direct2D
* For those who don't like to manage widgets (aka controls, aka stuff the user clicks on), NWG is for you. The UI manages the controls and the resources.
* For those who like documentation, NWG has one ( and I think it's pretty good ). Oh and its API is available online too: https://gabdube.github.io/native-windows-gui/ 
* For those who want a light and simple API, NWG might be for you

And it isn't...

* For those who want portability across system. Maybe it will work with WINE though...
* For those who want to deploy a production ready application as soon as possible. The first stable version will take some time to come out.
* For those who want a safe api to create custom control, nwg is not there YET...
* For those who want a UI to track a killers API address, Visual Basic is better (source: CSI)

# License

NWG use the MIT license

# Installation
To use NWG in your project add it to cargo.toml: 

```toml
[dependencies]
native-windows-gui = "0.2.0"
```

And then, in main.rs or lib.rs : 

```rust
extern crate native_windows_gui as nwg;
```

# Documentation

NWG has a complete documentation available here:  https://gabdube.github.io/native-windows-gui/

The documentation alone should be enough to introduce to the basics of NWG.

(btw) If English is your first language (it's not mine), it would be super kind to give me feedback about quality of the docs.

# Example
Having cargo installed and in your PATH, execute the following code to run the included example:  
(scroll further down for a code example)

```bash
git clone git@github.com:gabdube/native-windows-gui.git
cd native-windows-gui
cargo run --example showcase
cargo run --example dynamics
cargo run --example templating
cargo run --example simple_canvas --features=canvas
```

![A GUI](/img/showcase.png "Image")  

![A GUI](/img/canvas.png "Image")  

# Code Example

```rust
/**
    Simple example on how to use the nwg template system.
*/

#[macro_use] extern crate native_windows_gui as nwg;

//To hide console window
#![windows_subsystem = "windows"]

use nwg::{Ui, simple_message, fatal_message, dispatch_events};
use nwg::events as nwge;

nwg_template!(
    head: setup_ui<&'static str>,
    controls: [
        ("MainWindow", nwg_window!( title="Template Example"; size=(280, 105) )),
        
        ("Label1", nwg_label!( 
           parent="MainWindow"; text="Your Name: ";
           position=(5,15); size=(80, 25); font=Some("TextFont") )),
        
        ("YourName", nwg_textinput!( 
           parent="MainWindow"; position=(85,13); 
           size=(185,22); font=Some("TextFont") )),
        
        ("HelloButton", nwg_button!( 
           parent="MainWindow"; text="Hello World!";
           position=(5, 45); size=(270, 50); font=Some("MainFont") ))
    ];
    events: [
        ("HelloButton", "SaySomething", nwge::button::Click, |ui,_,_,_| {
            let your_name = nwg_get!(ui; ("YourName", nwg::TextInput));
            simple_message("Hello", &format!("Hello {}!", your_name.get_text()) );
        })
    ];
    resources: [
        ("MainFont", nwg_font!(family="Arial"; size=27)),
        ("TextFont", nwg_font!(family="Arial"; size=17))
    ];
    values: []
);

fn main() {
    let app: Ui<&'static str>;

    match Ui::new() {
        Ok(_app) => { app = _app; },
        Err(e) => { fatal_message("Fatal Error", &format!("{:?}", e) ); }
    }

    if let Err(e) = setup_ui(&app) {
        fatal_message("Fatal Error", &format!("{:?}", e));
    }

    dispatch_events();
}
```


CHANGELOG.md
# BETA 2 (0.3.0)

## Announcement

* License switch: GPLv3 to MIT 
* It is now possible to compile NWG on the gnu toolchain

## Breaking changes

* New way to handle events. Allow custom event definition and improve the event dispatching loop.  
For example, instead of `nwg::Event::Click`, use `nwg::events::button::Click`. 
For more information also see: 
  * https://gabdube.github.io/native-windows-gui/book/events.html (events basics)
  * https://gabdube.github.io/native-windows-gui/book/custom_events.html (defining custom events)
  * https://github.com/gabdube/native-windows-gui/blob/master/examples/templating.rs (simple example)
* Events are no longer restrained on controls. This means that it is now possible to extend builtin controls with custom user events!
* Uis no longuer implictly free the children when unpacking a control. Instead the children handles
  must be returned in a `Vec` by the `Control.children` method. This method can be ignored if the
  control can't have children.
* Reworked the `Canvas` control internals. Now the canvas resource are simple NWG resources.
For more information also see: 
  * https://gabdube.github.io/native-windows-gui/book/canvas.html (canvas basics)
* The `Canvas` control is now feature gated behind the feature `canvas`. This is because `d2d1.lib` is not included with the gnu version
  and requires a few extra (annoying) steps. 

## New resources and controls

* The **Image** resource to load bitmap, ico and cursor files
* The **ImageFrame** control to display a bitmap in a window
* The **Frame** control to display a bordered frame inside another window
* The **TreeView** control. To display tree hierarchy of data
* The **TreeViewItem** control. To display an item in a tree view control
* The **ContextMenu** control. A pop-up menu that can be shown anywhere in screen. usually pops when the user right click the mouse.
* The **TabView** control. A container that display Tab controls
* The **Tab** control. A special window that can be added to a TabView

## Existsing control changes

* **Window**
    * `set_icon` and `get_icon`: Allow the user to set or get the window icon  

* **UI**
  * `has_handle`: Check if the ui has an object identified by an handle
  * `id_from_handle`: Return the `ID` associated with an HANDLE
  * `type_of_control`: Return the `ControlType` associated with a control

* **Groupbox**
  * Now implemented as a custom control. Fix flicking issue and black paddings.

* **Most controls**
  * `set_font` and `get_font`: Allow the user to set the font or get the font identifier of many built-in controls  
  * `update`: Force a control to redraw itself.  
  * `focus`: Sets the keyboard focus to the specified control


## Other methods

* `toggle_console`: Hide or show the program console
* The `Cursor` struct. A fieldless struct to interface over the system cursor.
    * `get_position`: Return the cursor position in the screen
    * `set_position`: Set the cursor position in the screen
    * `get`: Set the cursor look (using a Cursor Image resource)
    * `set`: Get the cursor resource identifier (using a Cursor Image resource)
    * `get_capture`: Get the identifier of the control that captures the mouse
    * `set_capture`: Capture the mouse for a control 
    * `release`: Release the mouse capture set with `set_capture`
    * `dragging`: Check if the user tries to drag a control. Mostly used with a `set_capture`/`release` pair   


# BETA 1 (0.2.0)

Initial beta release. This is a complete rewrite of the ALPHA, so there's nothing to list

examples/showcase.rs
/**
    Example that show every control implemented in NWG
*/

#[macro_use] extern crate native_windows_gui as nwg;

use nwg::{Ui, EventArgs, dispatch_events, exit as nwg_exit};
use nwg::constants::{FONT_WEIGHT_BLACK, FONT_DECO_ITALIC, CheckState, FileDialogAction, HTextAlign, PickerDate, ImageType};
use nwg::events as nwge;    

static EMBED_BMP01: &'static [u8; 27702]  = include_bytes!("../img/rust-logo.bmp");

nwg_template!(
    head: setup_ui<&'static str>,
    controls: [
        // Window
        ("MainWindow", nwg_window!(title="Nwg Showcase"; position=(100, 100); size=(500, 400); icon=Some("RustLogoIcon"))),
        
        // Menus
        ("FileMenu", nwg_menu!(parent="MainWindow"; text="&File")),
        ("TestSubmenu1", nwg_menu!(parent="FileMenu"; text="&Submenu")),
        ("TestDisabledSubmenu", nwg_menu!(parent="FileMenu"; text="Disabled Submenu"; disabled=true)),
        ("TestSubmenu2", nwg_menu!(parent="TestSubmenu1"; text="&Another submenu")),
        ("NestedAction", nwg_menuitem!(parent="TestSubmenu2"; text="H&ello")),
        ("DisabledNestedAction", nwg_menuitem!(parent="TestSubmenu2"; text="Disabled"; disabled=true)),
        ("S1", nwg_separator!(parent="FileMenu")),
        ("QuitItem", nwg_menuitem!(parent="FileMenu"; text="&Quit")),
        ("WindowAction", nwg_menuitem!(parent="MainWindow"; text="&Action")),

        // Context Menu
        ("Context", nwg_contextmenu!()),
        ("Action1", nwg_menuitem!(parent="Context"; text="Action 1")),
        ("Action2", nwg_menuitem!(parent="Context"; text="Action 2")),
        ("S2", nwg_separator!(parent="Context")),
        ("TestSubmenu3", nwg_menu!(parent="Context"; text="&Submenu")),
        ("Action3", nwg_menuitem!(parent="TestSubmenu3"; text="SayHello")),
        ("Action4", nwg_menuitem!(parent="TestSubmenu3"; text="Disabled :("; disabled=true)),
        
        // Tabs
        ("TabView", nwg_tabsview!(parent="MainWindow"; position=(5, 5); size=(490, 370); font=Some("Font1"))),
        ("Tab1", nwg_tab!(parent="TabView"; text="Simple controls")),
        ("Tab2", nwg_tab!(parent="TabView"; text="Images And Trees")),
        ("Tab3", nwg_tab!(parent="TabView"; text="Listview")),

        // Timers
        ("TimerButton", nwg_button!(parent="Tab1"; text="Start timer"; position=(10,5); size=(100, 30); font=Some("Font2"))),
        ("TimerLabel", nwg_label!(parent="Tab1"; text="Time elapsed: 0 seconds"; position=(120, 10); size=(200, 25); font=Some("Font2"))),
        ("Timer", nwg_timer!(interval=500)),

        // Combobox
        ("SchoolSupplyComboBox ", 
            nwg_combobox!(
                parent="Tab1";
                position=(325, 10); size=(145, 30);
                placeholder=Some("Choose plz"); font=Some("Font1");
                collection=vec!["Pencil", "Eraser", "Scissor", "Calculator", "Notebook"])),

        // File dialog
        ("FileDialogButton", nwg_button!(parent="Tab1"; text="Browse File"; position=(10,40); size=(100, 30); font=Some("Font1"))),
        ("FilePathInput", nwg_textinput!(parent="Tab1"; position=(120, 45); size=(350, 24); readonly=true; font=Some("Font1"))),
        ("FileDialog", nwg_filedialog!(parent=Some("MainWindow"); action=FileDialogAction::Open; filters=Some("Test(*.txt;*.rs)|Any(*.*)"))),

        // List
        ("NameList", nwg_listbox!(
            parent="Tab1"; 
            position=(10, 75); size=(150, 100);
            collection=vec!["A Listbox", "Jimmy", "Sam", "Coconut", "Waldo", "David", "John"])),
        
        // Checkbox & radios
        ("HappyCheckBox", nwg_checkbox!(parent="Tab1"; text="I am happy"; position=(170, 75); size=(110, 30); checkstate=CheckState::Checked; font=Some("Font1"))),
        ("TriCheckBox", nwg_checkbox!(parent="Tab1"; text="Three states"; position=(290, 75); size=(110, 30); tristate=true; checkstate=CheckState::Indeterminate; font=Some("Font1"))),
        ("CatRadio", nwg_radiobutton!(parent="Tab1"; text="I have a cat"; position=(170, 105); size=(110, 30); checkstate=CheckState::Checked; font=Some("Font1"))),
        ("DogRadio", nwg_radiobutton!(parent="Tab1"; text="I have a dog"; position=(290, 105); size=(110, 30); font=Some("Font1"))),
        
        // Groupbox
        ("YesNoGroup", nwg_groupbox!(parent="Tab1"; text="Choose one"; position=(10, 185); size=(150, 60);  align=HTextAlign::Center; font=Some("Font1") )),
        ("YesRadio", nwg_radiobutton!(parent="YesNoGroup"; text="Blue"; position=(10, 20); size=(50, 30); font=Some("Font1"))),
        ("NoRadio", nwg_radiobutton!(parent="YesNoGroup"; text="Red"; position=(90, 20); size=(50, 30); font=Some("Font1"))),
        
        // Textbox
        ("RandomStuffLabel", nwg_label!(parent="Tab1"; text="Write your notes here:"; position=(170, 135); size=(180, 25); font=Some("Font1"))),
        ("RandomStuffTextBox", nwg_textbox!(parent="Tab1"; position=(170, 160); size=(290, 100); scrollbars=(false, true))),
        
        // Progress bar
        ("InstallCatLabel", nwg_label!(parent="Tab1"; text="Installing cat.exe ..."; position=(10, 265); size=(150, 25); font=Some("Font1") )),
        ("CatProgress", nwg_progressbar!(parent="Tab1"; position=(10, 290); size=(200, 25); range=(0, 100); value=85)),
        
        // Date picker
        ("DateLabel", nwg_label!(parent="Tab1"; text="Select the birth date of your cat"; position=(220, 265); size=(220, 25); font=Some("Font1") )),
        ("DatePicker", nwg_datepicker!(parent="Tab1"; value=Some(PickerDate{year:2016, month:12, day:1}); format=" dd MMMM yyyy"; position=(220, 290); size=(240, 25); font=Some("Font1"))),
        
        // ImageFrame
        ("RustLogoFrame", nwg_image_frame!(parent="Tab2"; image=Some("RustLogo"); position=(190, 215); size=(100,100))),
        ("OtherFrame", nwg_image_frame!(parent="Tab2"; image=Some("RustLogoMemory"); position=(295, 215); size=(100,100))),
        
        // TreeView
        ("TreeSelected", nwg_textinput!(parent="Tab2"; position=(190, 5); size=(280, 22); placeholder=Some("Selected Item Text"); font=Some("Font1") )),
        ("TreeView", nwg_treeview!(parent="Tab2"; position=(5, 5); size=(180, 315))),
        ("Tree_Root", nwg_treeview_item!(parent="TreeView"; text="Department")),
        ("TreeDirector", nwg_treeview_item!(parent="Tree_Root"; text="Director & Associate")),
        ("TreeBob", nwg_treeview_item!(parent="TreeDirector"; text="Bob Stalone")),
        ("TreeJob", nwg_treeview_item!(parent="TreeDirector"; text="Job Drake")),
        ("TreeManagement", nwg_treeview_item!(parent="Tree_Root"; text="Management")),
        ("TreeSally", nwg_treeview_item!(parent="TreeManagement"; text="Sally Foo")),
        ("TreeTI", nwg_treeview_item!(parent="Tree_Root"; text="TI")),

        // Listview
        ("ListView", nwg_listview!(parent="Tab3"; position=(5, 5); size=(280, 315); font=Some("Font1")))
    ];
    events: [
        ("RandomStuffTextBox", "AllSystemEvents", nwge::Any, |_,_,_,args| {
            let (msg, w) = match args {
                &EventArgs::Raw(msg, w, _) => (msg, w),
                _ => unreachable!()
            };

            if msg == 0x0100 { // WM_KEYDOWN
                println!("The virtual key {:?} was pressed", w);
            }
        }),

        ("TreeView", "ItemSelected", nwge::treeview::ItemChanged, |app,_,_,_| {
            let (tree, text) = nwg_get!(app; [
                ("TreeView", nwg::TreeView),
                ("TreeSelected", nwg::TextInput)
            ]);

            if let Some(item_id) = tree.get_selected_item(app) {
                let item = nwg_get!(app; (item_id, nwg::TreeViewItem));
                text.set_text( &item.get_text() );
            }
        }),

        ("TimerButton", "Start Timer", nwge::button::Click, |app,_,_,_|{
            let (mut timer, btn) = nwg_get_mut!(app; [
                ("Timer", nwg::Timer),
                ("TimerButton", nwg::Button)
            ]);

            if timer.running() {
                btn.set_text("Start Timer");
                timer.stop();
            } else {
                btn.set_text("Stop Timer");
                timer.start();
            }
        }),

        ("Timer", "UpdateLabel", nwge::timer::Tick, |app,_,_,_|{
            let (label, timer) = nwg_get!(app; [("TimerLabel", nwg::Label), ("Timer", nwg::Timer)]);
            label.set_text(format!("Time elapsed: {:?} seconds", timer.elapsed().as_secs()).as_ref());
        }),

        ("FileDialogButton", "ChooseFile", nwge::button::Click, |app,_,_,_|{
            let (dialog, file_path) = nwg_get_mut!(app; [
                ("FileDialog", nwg::FileDialog),
                ("FilePathInput", nwg::TextInput)
            ]);

            if dialog.run() {
                file_path.set_text(&dialog.get_selected_item().unwrap());
            }
        }),

        ("DatePicker", "DD", nwge::datepicker::DateChanged, |app,_,_,_| {
            println!("{:?}", nwg_get!(app; ("DatePicker", nwg::DatePicker)).get_value());
        }),

        ("RustLogoFrame", "Logo", nwge::image_frame::Click, |app,_,_,_| {
            let img = nwg_get!(app; ("RustLogoFrame", nwg::ImageFrame));
            if let Some("RustLogo") = img.get_image(app) {
                img.set_image(app, Some(&"RustMascot")).unwrap();
            } else {
                img.set_image(app, Some(&"RustLogo")).unwrap();
            }
        }),

        ("MainWindow", "context", nwge::MouseDown, |app, _, _, args| {
            match args {
                &EventArgs::MouseClick{ref btn, pos: _} => {
                    if let nwg::constants::MouseButton::Right = *btn {
                        // WATCH OUT! The position param of MouseClick use local coordinate, but the `pop_at` method requires global coordinates
                        // To make sure the menu appears at the right place, always use the `Cursor::get_position` method.
                        let (x, y) = nwg::Cursor::get_position();   
                        nwg_get!(app; ("Context", nwg::ContextMenu)).pop_at(x, y);
                    }
                },
                _ => unreachable!()
            }
        }),

        ("NestedAction", "SayHello", nwge::menu::Triggered, |_,_,_,_| { nwg::simple_message("Hello", "Hello World!");  }),
        ("Action3", "SayHello", nwge::menu::Triggered, |_,_,_,_| { nwg::simple_message("Hello", "Hello World!"); }),
        ("QuitItem", "Quit", nwge::menu::Triggered, |_,_,_,_| { nwg_exit() })

    ];
    resources: [
        ("Font1", nwg_font!(family="Calibri"; size=17 )),
        ("Font2", nwg_font!(family="Arial"; size=17; weight=FONT_WEIGHT_BLACK; decoration=FONT_DECO_ITALIC)),
        ("RustLogo", nwg_image!(source="img\\rust-logo.bmp"; image_type=ImageType::Bitmap)), // Make sure to use '\\' and not '/'
        ("RustMascot", nwg_image!(source="img\\rust-mascot.bmp"; image_type=ImageType::Bitmap; size=(100, 100))), // Make sure to use '\\' and not '/'
        ("RustLogoIcon", nwg_image!(source="img\\rust-logo.ico"; image_type=ImageType::Icon)),
        ("RustLogoMemory", nwg::MemoryImageT{source: Vec::from( &EMBED_BMP01[..] )} )
    ];
    values: []
);

pub fn main() {
    let app: Ui<&'static str> = Ui::new().expect("Failed to initialize the Ui");
    
    // Pack the control in the application
    if let Err(e) = setup_ui(&app) {
        panic!("Commit failed: {:?}", e);
    }

    // Dispatch the events until the user quits
    dispatch_events();
}